Move generate code into a separate library

hmeng-19 opened the pull request:
Move generate code into a separate library, `libgen`. 

To avoid using `cli.Context` inside `libgen`, the `modify` function responds to get the command options, and transfer them into the functions inside `libgen`.

The PR tries to fix #130 .

hmeng-19 commented:
@mrunalp , PTAL.

hmeng-19 commented:
@mrunalp , @wking , this PR only aims to move the generate code into a separate library.  
We may do another PR to coordinate with #128 to make reviewing PRs easier.

wking commented on generate.go:
```diff
@@ -9,9 +9,8 @@ import (
 	"strings"
 
-	"github.com/Sirupsen/logrus"
+	"github.com/opencontainers/ocitools/libgen"
```
There was a reasonable amount of “what should we call this?” discussion in #128, which resolved around [`seccomp-gen`][1].  If we're aiming for separate generate/validate/([test][1][?][2]) commands, we probably want something like:

    ocitools
    |-- cmd
    |    `-- ocitools
    |         |-- main.go
    |         |-- generate.go
    |         |-- validate.go
    |         …
    |-- generate
    |    |-- generate.go
    |    |-- utils.go, etc.
    |    `-- seccomp (#128)
    |         |-- consts.go
    |         …
    |-- validate
    |    |-- validate.go
    |    `-- utils.go
    …

Shifting the import paths around while we settle into whatever layout we choose isn't a big deal as long as we don't cut releases.  But I expect that the libraries will start picking up external consumers shortly after they're born, and we'll make life easier on those downstream folks if we don't churn the layout too much ;).

[1]: https://github.com/opencontainers/ocitools/pull/128#issuecomment-230688108
[2]: https://github.com/opencontainers/ocitools/pull/47#issuecomment-215328958
[3]: https://github.com/opencontainers/ocitools/pull/61#issuecomment-217910342

hmeng-19 replied:
@wking , this new package structure looks very different from the current package structure.

    ocitools
    |-- cmd
    |    `-- runtimetest
    |         |-- main.go
    |         |-- rlimit_linux.go
    |         |-- mount
    |               |-- mountinfo_freebsd.go
    |               |-- mountinfo.go
    |               |-- mountinfo_linux.go
    |               |-- mountinfo_solaris.go
    |               |-- mountinfo_unsupported.go
    |               |-- mountinfo_windows.go
    |-- main.go
    |-- generate.go
    |-- validate.go

hmeng-19 replied:
@wking , @mrunalp , I think I will first get myself familiar with #54 , and #128 . 
Then we can come back to this PR and decide how we want it to be.

wking replied:
On Mon, Jul 11, 2016 at 07:09:30AM -0700, hmeng-19 wrote:
> > @@ -9,9 +9,8 @@ import (
> >  	"strconv"
> >  	"strings"
> >  
> > -	"github.com/Sirupsen/logrus"
> > +	"github.com/opencontainers/ocitools/libgen"
>
> @wking , this new package structure looks very different from the
> current package structure…

Yeah, but if we're going to restructure things, I think we want to
take a big step to get to where we (think we) want to be, instead of
taking a bunch of short steps.  The big step may not be perfect, but
it's got a better chance at being right than a small step after which
we already plan on making further changes.

mrunalp replied:
@wking proposed layout scheme looks reasonable to me.

hmeng-19 replied:
@wking , @mrunalp . Sounds good. I will modify the code structure as @wking  suggested.
@wking , in your proposal, there are two `generate.go` - one is under `cmd/ocitools`, another is under `generate`. What is their difference? The same problem happens to `validate.go`.
I think all the files under `generate`  and `validate` should be libraries, right?

wking replied:
On Mon, Jul 11, 2016 at 11:36:11AM -0700, hmeng-19 wrote:
> @wking , in your proposal, there are two `generate.go` - one is
> under `cmd/ocitools`, another is under `generate`. What is their
> difference? The same problem happens to `validate.go`.

The cmd/ocitools stuff is just thin bindings between the library code
(under generate/, validate/, etc.) and urfave/cli.

> I think all the files under `generate` and `validate` should be
> libraries, right?

Yup.

wking commented on libgen/utils.go:
```diff
@@ -0,0 +1,649 @@
+	defer cf.Close()
+
+	if err = json.NewDecoder(cf).Decode(&spec); err != nil {
+		return
```
I'm not sure how this compiles (although it seems to), but I expect it should be returning `nil, err`.

wking replied:
And I see that this is just a copy/paste from the old code, but we might as well fix it now ;).

mrunalp replied:
Named returns ;) But yes should be similar to other returns in the code for consistency.

wking commented:
On Fri, Jul 08, 2016 at 08:02:51PM -0700, hmeng-19 wrote:
> We may do another PR to coordinate with #128 to make reviewing PRs
> easier.

This is also going to create some tedious conflicts with #54 (now
stable for almost two months ;).  Can we get an up/down ruling on that
PR in the near future.  Having these major PRs all touching the same
code and in flight at the same time is just going to create busywork.

wking commented on libgen/utils.go:
```diff
@@ -0,0 +1,649 @@
+}
+
+// AddIDMappings add the mappings of uid and gid into spec.
+func AddIDMappings(spec *rspec.Spec, uidMaps, gidMaps []string) error {
```
This should probably be a single:

    func ParseIDMappings(spec *rspec.Spec, idType string, maps []string) error {…}

which the command implementation can call with something like (untested, and I don't do much Go ;):

    idTypes := ["uid", "gid"]
    for idType := range idTypes {
        maps := context.StringSlice(idType + "mappings")
        if err := libgen.AddIDMappings(spec, idType, maps) err != nil {
          return err
        }
     }

The approach your copy/pasting was acceptable for an internal helper, but it seems like we should aim for something DRYer for a library method.  On the other hand, I'm fine cleaning this stuff up in follow-up PRs, as long as we don't wait long enough for API-breaking to become a concern.

wking commented on libgen/utils.go:
```diff
@@ -0,0 +1,649 @@
+}
+
+// AddHooks add the (prestart|poststop|poststart)hooks into spec.
+func AddHooks(spec *rspec.Spec, preStartHooks, postStopHooks, postStartHooks []string) error {
```
I would like to DRY this up [too][1], but again, that can happen in a follow up PR.  On the other hand, shifting functions out into a separate package seems like a good time to put some thought into the API we want to expose.  Just moving things without cleaning up the APIs seems like it's breaking #128 and #54 without much benefit.

[1]: https://github.com/opencontainers/ocitools/pull/131/files#r70162128

hmeng-19 replied:
@wking , reconsidering the API seems a good idea.
For this PR, I was just trying to separate `generate.go` and the library, and tried to avoid drastic changes of the code.
I will change the code after the new API is decided.

wking commented on libgen/utils.go:
```diff
@@ -0,0 +1,649 @@
+func StrPtr(s string) *string { return &s }
+
+// GetDefaultTemplate returns the default spec.
+func GetDefaultTemplate() *rspec.Spec {
```
I'd put this in `default.go`.  And does Go have a deep copy?  If so, I'd consider making this a variable.  That way it's easy for folks to swap in an alternative default template or mutate the original if they want, and they can make a deep copy if they don't want to mutate the default template.

hmeng-19 replied:
@wking , I thought of putting this into `default.go`, and did not do it when I noticed that there are lots of maps, pointers, and slices embedded in the `rspec.Spec`.
I will try to see whether it is easy to do a deep copy of a `rspec.Spec` in Go.

mrunalp commented:
I agree with @wking that we should combine this with #128 so we have a single cohesive change/library.

hmeng-19 commented:
@mrunalp , I am okay with the plan. 
Should I also try to modify the code structure of the current `validate.go` file? If so, we probabaly should rename the title of #130 to be more generic.

hmeng-19 commented:
@mrunalp , @wking , the CI checks failed because this PR tries to introduce a new package `cmd/ocitools`:
 
    The command "$HOME/gopath/bin/git-validation -run DCO,short-subject -v -range ${TRAVIS_COMMIT_RANGE}" exited with 0.
    0.01s$ make
    go build -tags "" -o ocitools ./cmd/ocitools
    can't load package: package github.com/opencontainers/ocitools/cmd/ocitools: cannot find package "github.com/opencontainers/ocitools/cmd/ocitools" in any of:
        /home/travis/.gimme/versions/go1.5.3.linux.amd64/src/github.com/opencontainers/ocitools/cmd/ocitools (from $GOROOT)
        /home/travis/gopath/src/github.com/opencontainers/ocitools/Godeps/_workspace/src/github.com/opencontainers/ocitools/cmd/ocitools (from $GOPATH)
        /home/travis/gopath/src/github.com/opencontainers/ocitools/cmd/ocitools
    make: *** [all] Error 1

hmeng-19 commented:
@mrunalp , @wking , the new changes I made include:
 
    rename libgen -> generate
    move package main into cmd/ocitools
    Add an empty generate/seccomp package
  
The new changes focus on the package structure, does not really polish the function API. 
I think we can merge this PR to unblock #128. Then we can make a new PR to polish the new function API.

wking commented:
On Mon, Jul 11, 2016 at 02:37:44PM -0700, hmeng-19 wrote:
> @mrunalp , @wking , the CI checks failed because this PR tries to
> introduce a new package `cmd/ocitools`:
>  
> …
>     can't load package: package github.com/opencontainers/ocitools/cmd/ocitools: cannot find package "github.com/opencontainers/ocitools/cmd/ocitools" in any of:
>         /home/travis/.gimme/versions/go1.5.3.linux.amd64/src/github.com/opencontainers/ocitools/cmd/ocitools (from $GOROOT)
>         /home/travis/gopath/src/github.com/opencontainers/ocitools/Godeps/_workspace/src/github.com/opencontainers/ocitools/cmd/ocitools (from $GOPATH)
>         /home/travis/gopath/src/github.com/opencontainers/ocitools/cmd/ocitools

Actually, I think that's failing because you haven't added
cmd/ocitools [1].  Did you forget a ‘git add cmd/ocitools’ with
d2f10ede?

hmeng-19 commented:
@wking , you are right. My `.gitignore` confused me and I thought I added those files.
Fix it. Thanks. PTAL.

wking commented:
On Mon, Jul 11, 2016 at 02:55:12PM -0700, hmeng-19 wrote:
> My `.gitignore` confused me and I thought I added those files.

Ah, the ‘ocitools’ entry in .gitignore should probably be ‘/ocitools’
to anchor the ignore (and not ignore cmd/ocitools).  9777ceb made a
similar fix for runtimetest.

I'm not sure what the ‘oci’ entry in .gitignore is for.  It's from
7ac06868 (2015-09-15), and maybe is just stale?

hmeng-19 commented:
@wking , I think the problem is the pattern `ocitools` in the `.gitignore` file, which automatically
ignore all the files or directories with the name of `ocitools`. In our case, `cmd/ocitools`, together with all the entries under it, gets ignored.
However, by adding a leading slash into a pattern, it would only match the beginning of the pathname. That is how Mrunal solved the problem in 9777ceb473cf23005ac74f177d83fc7398845a7b. `/runtimetest` only matches `runtimetest` directly under the git repo.

Please refer to here https://git-scm.com/docs/gitignore .

hmeng-19 commented:
@wking , @mrunalp , fix the `.gitignore` file. PTAL.

wking commented:
Through fed976fa6 look fine to me if we want to shuffle things around
without changing APIs (much), although I'd rather it was squashed down
into a single commit.

But I don't see a need to split “restructure and reconsider the API”
into two separate PRs [1].  And I think it would be nice if #54 and
#128 (which were both filed before this change was proposed in #130)
got reviewed first.

[1]: https://github.com/opencontainers/ocitools/pull/131#discussion_r70261507

GrantSeltzer commented:
@hmeng-19 and I are in agreement that this PR should get merged before #128, and i'll probably close #128 and make a new one that will add all of the seccomp stuff into the generate library that she is separating out.

wking commented:
On Tue, Jul 12, 2016 at 07:13:44AM -0700, Grant Seltzer Richman wrote:
> @hmeng-19 and I are in agreement that this PR should get merged
> before #128, and i'll probably close #128 and make a new one that
> will add all of the seccomp stuff into the generate library that she
> is separating out.

Whatever's easier, but you can rebase your existing #128 branch once
this PR lands.  And if the plan is to land this PR first, I'll survive
a rebase of #54 (but it would be nice if PRs were reviewed and
merged/rejected more quickly to avoid these parallel-work conflicts.

mrunalp commented on cmd/ocitools/generate.go:
```diff
@@ -0,0 +1,244 @@
+	Flags:  generateFlags,
+	Before: before,
+	Action: func(context *cli.Context) error {
+		spec := generate.GetDefaultTemplate()
```
This could be `generate.New()`

mrunalp replied:
Again, a rename suggestion for `GetDefaultTemplate` to `New`

mrunalp commented on cmd/ocitools/generate.go:
```diff
@@ -0,0 +1,244 @@
+		template := context.String("template")
+		if template != "" {
+			var err error
+			spec, err = generate.LoadTemplate(template)
```
`generate.NewFromTemplate()`

hmeng-19 replied:
@mrunalp , are you suggesting move the following code into a new func `generate.NewFromTemplate`?

    template := context.String("template")
    if template != "" {
        var err error
        spec, err = generate.LoadTemplate(template)

hmeng-19 replied:
@mrunalp , should we name it `NewSpecFromTemplate`?

mrunalp replied:
Yeah, I am suggesting that we rename it.

mrunalp commented:
commit history needs to be cleaned up or they should be squashed.

hmeng-19 commented:
@mrunalp , @wking , renamed the two functions you suggested, and squashed all my changes into a single commit. 
PTAL.

mrunalp commented:
While this is a step in the right direction,  I think API like the following will be easier to use:

```go

type generator struct {
    spec *rspec.Spec
}

// Start from default template
specgen := generator.New()

// Make modifications
specgen.SetRootPath("/path/to/rootfs")
specgen.AddCapabilities(...)
specgen.AddTmpfsMount(...)
specgen.SetSelinuxLabel(...)
...
...
// Write to disk
specgen.Save("/path/to/file")

spec := specgen.GetSpec()

// Load a template.json
specgen := generator.NewFromTemplate("/path/to/file")
specgen.AddMount(src, dest, options..)
specgen.Save("/some/other/path")

// Load existing spec object and modify it to generate a new spec
specgen := generator.NewFromSpec(someSpecObject)
specgen.SetNoNewPriviliges()
specgen.Save("/yet/another/path")
```

wking commented:
On Wed, Jul 13, 2016 at 02:53:54PM -0700, Mrunal Patel wrote:
> specgen := generator.NewFromTemplate("/path/to/file")

I'd load this from an io.Reader for more flexibility.  Wrapping that
in something that opened a file, deferred a close, and passed the
writer in to generator.NewFromReader is acceptable noise for the
command-line binding package.

mrunalp commented:
Also, we should have both Set and Add for items that are lists (like Capabilities and Namespaces for e.g.). That allows one to reset the list or set it wholesale using Set and add items to it using Add. Drop should also be added for Capabilities and Namespaces to begin with.

hmeng-19 commented:
@mrunalp , @wking , sounds good.

hmeng-19 commented:
@mrunalp , @wking , PTAL.
I rebased the PR on top of the latest master. 
All the new change since the last review is in the second commit https://github.com/opencontainers/ocitools/pull/131/commits/5b4ed54f15dc2f39d396d36a67f35aaaa7d12490 .

hmeng-19 commented:
@mrunalp , @wking ,  ignore the last message.
I still missed some `Add` function. Will fix it soon.

mrunalp commented on cmd/ocitools/generate.go:
```diff
@@ -0,0 +1,299 @@
+			}
+		}
+
+		err := modify(specgen, context)
```
We should rename `modify` to `setupSpec` or something similar.

hmeng-19 replied:
Sounds good. Work on it.

mrunalp commented on cmd/ocitools/generate.go:
```diff
@@ -0,0 +1,299 @@
+	if len(spec.Version) == 0 {
+		spec.Version = rspec.Version
+	}
+	spec.Root.Path = context.String("rootfs")
```
The code from here through line 127 should all have functions in the generator.

hmeng-19 replied:
@mrunalp , Okay. In general, we can provide get/set (plus add if applicable) functions for each field inside g.spec. Right?

mrunalp replied:
We don't need get except for spec

Sent from my iPhone

> On Jul 15, 2016, at 7:12 AM, hmeng-19 <notifications@github.com> wrote:
> 
> In cmd/ocitools/generate.go:
> 
> > +			err = specgen.Save(os.Stdout)
> > +		}
> > +		if err != nil {
> > +			return err
> > +		}
> > +		return nil
> > +	},
> > +}
> > +
> > +func modify(g generate.Generator, context *cli.Context) error {
> > +	spec := g.GetSpec()
> > +
> > +	if len(spec.Version) == 0 {
> > +		spec.Version = rspec.Version
> > +	}
> > +	spec.Root.Path = context.String("rootfs")
> @mrunalp , Okay. In general, we can provide get/set (plus add if applicable) functions for each field inside g.spec. Right?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

hmeng-19 replied:
Yes. I think you are right. Realized it when I am fixing it.

hmeng-19 commented:
@mrunalp , just let know. I am still trying to polishing each function. The current commit is not the final one.

mrunalp commented on cmd/ocitools/generate.go:
```diff
@@ -0,0 +1,345 @@
+	if context.IsSet("privileged") {
+		privileged = context.Bool("privileged")
+	}
+	g.SetupPrivilege(privileged)
```
`SetupPrivilege` -> `SetPrivileged`

hmeng-19 replied:
Accept.

mrunalp commented on cmd/ocitools/generate.go:
```diff
@@ -0,0 +1,345 @@
+
+	if context.IsSet("cap-add") {
+		addCaps := context.StringSlice("cap-add")
+		if err := g.AddProcessCapabilities(addCaps); err != nil {
```
`AddProcessCapabilities` -> `AddProcessCapability`

hmeng-19 replied:
Accept

mrunalp commented on cmd/ocitools/generate.go:
```diff
@@ -0,0 +1,345 @@
+
+	if context.IsSet("bind") {
+		binds := context.StringSlice("bind")
+		if err := g.AddBindMounts(binds); err != nil {
```
API should add one bind mount at a time and should be called AddBindMount

hmeng-19 replied:
Is it okay we expect the user to put a single bind mount setting into a slice?
Another I can do is use variadic function.

mrunalp replied:
We can loop in the calling code to add multiple bind mounts.

hmeng-19 replied:
Sounds good.

mrunalp commented on cmd/ocitools/generate.go:
```diff
@@ -0,0 +1,345 @@
+
+	if context.IsSet("prestart") {
+		preStartHooks := context.StringSlice("prestart")
+		if err := g.AddPreStartHooks(preStartHooks); err != nil {
```
Same here; one at a time.

mrunalp commented on cmd/ocitools/generate.go:
```diff
@@ -0,0 +1,345 @@
+
+	if context.IsSet("tmpfs") {
+		tmpfsSlice := context.StringSlice("tmpfs")
+		if err := g.AddTmpfsMounts(tmpfsSlice); err != nil {
```
Same here. This should be `AddTmpfsMount`

mrunalp commented on cmd/ocitools/generate.go:
```diff
@@ -0,0 +1,345 @@
+
+	if context.IsSet("poststop") {
+		postStopHooks := context.StringSlice("poststop")
+		if err := g.AddPostStopHooks(postStopHooks); err != nil {
```
Same for hooks.

hmeng-19 commented:
@mrunalp , I am fixing all these problems. The new commit is coming soon.

mrunalp commented on generate/utils.go:
```diff
@@ -0,0 +1,1147 @@
+// Package generate implements functions generating container config files.
```
rename this file to generate.go?

hmeng-19 replied:
Okay.

mrunalp commented on generate/utils.go:
```diff
@@ -0,0 +1,1147 @@
+}
+
+// StrPtr returns the pointer pointing to the string s.
+func StrPtr(s string) *string { return &s }
```
this function could be private to the generate package.

hmeng-19 replied:
The validate.go is currently using this function once.

wking replied:
On Fri, Jul 15, 2016 at 03:46:51PM -0700, hmeng-19 wrote:
> > +// StrPtr returns the pointer pointing to the string s.
> > +func StrPtr(s string) *string { return &s }
> 
> The validate.go is currently using this function once.

ocitools/util.go ;).

hmeng-19 replied:
Sounds good. Fixing it.

hmeng-19 replied:
I may remember the wrong variable. `strPtr` is only used by the generate library and can be private.

hmeng-19 commented:
@mrunalp , @wking , I am finishing my change on my side for now, and also rebase and squash my commit.
Do you want to push my changes to this PR now? or you guys want to finish review the current version?

mrunalp commented:
Please push your updates so the latest version could be reviewed.

Sent from my iPhone

> On Jul 15, 2016, at 3:58 PM, hmeng-19 <notifications@github.com> wrote:
> 
> @mrunalp , @wking , I am finishing my change on my side for now, and also rebase and squash my commit.
> Do you want to push my changes to this PR now? or you guys want to finish review the current version?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

hmeng-19 commented:
@mrunalp , @wking , the latest PR is in.

mrunalp commented on generate/generate.go:
```diff
@@ -0,0 +1,1015 @@
+}
+
+// ClearLinuxUIDMapping clears g.spec.Linux.UIDMappings.
+func (g Generator) ClearLinuxUIDMapping() {
```
ClearLinuxUIDmappings

mrunalp commented on generate/generate.go:
```diff
@@ -0,0 +1,1015 @@
+}
+
+// ClearLinuxGIDMapping clears g.spec.Linux.GIDMappings.
+func (g Generator) ClearLinuxGIDMapping() {
```
ClearLinuxGIDMappings

mrunalp commented on generate/generate.go:
```diff
@@ -0,0 +1,1015 @@
+}
+
+// ClearLinuxNamespace clears g.spec.Linux.Namespaces.
+func (g Generator) ClearLinuxNamespace() {
```
ClearLinuxNamespaces

mrunalp commented on generate/generate.go:
```diff
@@ -0,0 +1,1015 @@
+}
+
+// ClearPreStartHook clears g.spec.Hooks.Prestart.
+func (g Generator) ClearPreStartHook() {
```
ClearPreStartHooks

mrunalp commented on generate/generate.go:
```diff
@@ -0,0 +1,1015 @@
+				},
+			},
+		},
+		Hostname: "shell",
```
This should be something else. Maybe `example`.

wking replied:
On Fri, Jul 15, 2016 at 04:25:19PM -0700, Mrunal Patel wrote:
> > +		Hostname: "shell",
> 
> This should be something else. Maybe `example`.

I usually echo the spec's example [1].

[1]: https://github.com/opencontainers/runtime-spec/blame/v1.0.0-rc1/config.md#L188

hmeng-19 commented:
@mrunalp , @wking , fixed the problems you pointed out in a separate commit https://github.com/opencontainers/ocitools/pull/131/commits/0e6dfc00c26a3ca89008b81eb08684b370908ad7 . PTAL.

liangchenye commented:
0e6dfc0  LGTM

mrunalp commented on generate/generate.go:
```diff
@@ -0,0 +1,1015 @@
+
+func checkCap(c string) error {
+	isValid := false
+	cp := fmt.Sprintf("CAP_%s", c)
```
This isn't required. The list you are checking against doesn't have CAP_

hmeng-19 replied:
The list I am checking against is `capability.List()`, which includes `CAP_CHOWN`, `CAP_DAC_OVERRIDE`, and so on.

hmeng-19 replied:
Ignore my comment. I got your point.

mrunalp commented on generate/generate.go:
```diff
@@ -0,0 +1,1015 @@
+	cp := fmt.Sprintf("CAP_%s", c)
+
+	for _, cap := range capability.List() {
+		if strings.ToUpper(cp) == strings.ToUpper(cap.String()) {
```
Move this out of the loop to avoid calling ToUpper in each iteration for cp

mrunalp commented:
--seccomp-default doesn't get populated
```
[root@dhcp-16-129 busybox]# ocitools generate --tty --output=config.json --seccomp-default SCMP_ACT_ALLOW --seccomp-errno getcwd
[root@dhcp-16-129 busybox]# tail -n 20 config.json 
                        {
                                "type": "mount"
                        }
                ],
                "seccomp": {
                        "defaultAction": "",
                        "architectures": null,
                        "syscalls": [
                                {
                                        "name": "getcwd",
                                        "action": "SCMP_ACT_ERRNO"
                                }
                        ]
                }
        },
        "solaris": {
                "cappedCPU": {},
                "cappedMemory": {}
        }
```

mrunalp commented:
Also trying to add a capability or dropping it fails
Try --cap-add SYS_ADMIN. I have pointed to the problem in earlier comment.

hmeng-19 commented:
@mrunalp , fixed the problems you pointed out and also tested all the other options, PTAL.

mrunalp commented:
Can you re-order or squash the commits?

hmeng-19 commented:
@mrunalp , no problem. Are you done checking the latest changes?

mrunalp commented:
@hmeng-19 Yeah, the changes look good and am ready to merge.

hmeng-19 commented:
@mrunalp , I sqashed the commits.

mrunalp commented:
LGTM
